### PR TITLE
Add external services repos table

### PR DESCRIPTION
**Pull request by asdine**
The original PR #12789 failed to be deployed because of a migration error that was caused because of a deadlock (see https://sourcegraph.slack.com/archives/CMBA8F926/p1597923638000300 for more context).

This PR changes several things to the migration, these changes can be seen in this commit https://github.com/sourcegraph/sourcegraph/commit/7481ec8a8e7ba1ebd83a5b9927125c58e666e1cc:
- Acquire an exclusive table lock before the migration
- Make the sources column read-only, using a trigger, instead of removing it.
- Run the migration code only on non deleted repos

The script was tested on production, with a rollback instead of a commit, and it took 1m25s

### CODE REVIEW

File: migrations/1528395705_add_table_external_services_repos.up.sql [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    AS $$
+    BEGIN
+        IF (OLD.sources != NEW.sources) THEN
+            RAISE EXCEPTION 'sources is read-only';
+        END IF;
+
+        RETURN OLD;
+    END;
+$$;
+CREATE TRIGGER trig_read_only_repo_sources_column BEFORE UPDATE OF sources ON repo FOR EACH ROW EXECUTE PROCEDURE make_repo_sources_column_read_only();
```
Comment (ryanslade) [inline migrations/1528395705_add_table_external_services_repos.up.sql:25]: You should drop this in the down migration

**Review by ryanslade — APPROVED** [review]
LGTM apart from needing to update the down script

File: migrations/1528395705_add_table_external_services_repos.up.sql (hunk 1 above)
Comment (asdine) [reply migrations/1528395705_add_table_external_services_repos.up.sql:25]: Good catch

**Comment by unknwon** [comment]
I don't have context on yesterday's debugging session, so please don't wait on my review. I think approval from @ryanslade is sufficient.

**Pull request merged**